Review2

Shafagh-Sepehr opened the pull request:
second review

Shafagh-Sepehr committed:
add

Shafagh-Sepehr committed:
move hard coded path to a newly created appsettings.json file.

Shafagh-Sepehr committed:
create new exception for null or whitespace input.

Shafagh-Sepehr committed:
create UserOutput class for output related actions.

Shafagh-Sepehr committed:
create UserInput class for input related actions.

Shafagh-Sepehr committed:
refactor UserInput and UserOutput classes name and methods and also extract interface.

Shafagh-Sepehr committed:
replace appsettings.json with a public static class for better accessibility across the project.

Shafagh-Sepehr committed:
move hardcoded banned words to the AppSettings class.

Shafagh-Sepehr committed:
refactor StringToWordsProcessor's TrimSplitAndStemString function:
extract two methods from it
get the stemmer via constructor instead of this method

Shafagh-Sepehr committed:
extract methods from TrimSplitAndStemString method

Shafagh-Sepehr committed:
change folder structure

Shafagh-Sepehr committed:
move InvertedIndex folder to a new Application folder.

Shafagh-Sepehr committed:
extract words processing from QuerySearcher.cs

Shafagh-Sepehr committed:
change QueryProcessor's name to WordProcessor

Shafagh-Sepehr committed:
refactored Query searcher.
It now inputs the inverted index through constructor.

Shafagh-Sepehr committed:
extract class from QuerySearcher.cs
GetDocumentForXQuery methods are now each a class and there's a master DocumentReader class that encapsulates them.

Shafagh-Sepehr committed:
extract methods from QuerySearcher's Search method

Shafagh-Sepehr committed:
improve inject IDocumentReader to QuerySearcher class for future test mocks.

Shafagh-Sepehr committed:
rename WordProcessor folder to WordProcessors because a class with the same name exists in it.

Shafagh-Sepehr committed:
fix the bug in QuerySearcher where Words are passed to document readers before initialization.

Shafagh-Sepehr committed:
extract methods from Search method of QuerySearcher class.

Shafagh-Sepehr committed:
rename Search methods to better ones.

Shafagh-Sepehr committed:
extract method from InvertedIndexDictionaryBuilder's FillInvertedIndexFromFile method.

Shafagh-Sepehr committed:
create builder class for InvertedIndexDictionary to decouple it from dependencies and also making instances easier.

Shafagh-Sepehr committed:
move regex patterns to AppSettings class

Shafagh-Sepehr committed:
Add Dependency Injection to the project for all classes.
create a Construct method to pass the needed parameters.
remove unused InvertedIndexBuilder.cs.
info: tests are down.

Shafagh-Sepehr committed:
Capitalize first letter of the Message in NullInputException.

Shafagh-Sepehr committed:
Add Assertions for classes that their Construct method should be called.

Shafagh-Sepehr committed:
extract Search methods to classes in QuerySearcher

Shafagh-Sepehr committed:
delete unused using statements.

Shafagh-Sepehr committed:
encapsulate class parameters

Shafagh-Sepehr committed:
remove interface postfix from interface namespaces.

Shafagh-Sepehr committed:
create InvertedIndexFactory for ease of use by both the user and developer.
extract Services to its own class.

Shafagh-Sepehr committed:
hide classes from user by changing them from public to internal.

navidnasyry commented on FullTextSearch/IO/ConsoleOutput.cs:
معمولا به پیاده‌سازی‌های اینترفیس ها میگن Service و مثلا توی یه پوشه به اسم Services میذارن

navidnasyry commented on FullTextSearch/Exceptions/NullInputException.cs:
درباره پیاده‌سازی‌ کانستراکتور‌های دیفالت Exception بخون. بهتره همه کانستراکتور‌هاش رو پیاده کنیم.

navidnasyry commented on FullTextSearch/AppSettings.cs:
```diff
@@ -0,0 +1,39 @@
+internal static partial class AppSettings
+{
+    public const string DocumentsPath = "/home/shafagh/Desktop/EnglishData";
+
```
اینو از کانفیگ بخونی بهتره. چرا؟

navidnasyry commented on FullTextSearch/Application/DocumentsReader/AndDocumentsReader.cs:
```diff
@@ -0,0 +1,22 @@
+        
+        var result = new HashSet<string>(andDocsList[0]);
+        
+        for (var i = 1; i < andDocsList.Count; i++)
```
use foreach
why?

navidnasyry replied:
what is the difference between for and foreach?

navidnasyry commented on FullTextSearch/Application/DocumentsReader/DocumentReader.cs:
```diff
@@ -0,0 +1,22 @@
+    INotDocumentsReader notDocumentsReader)
+    : IDocumentReader
+{
+    private IAndDocumentsReader AndDocumentsReader { get; } = andDocumentsReader;
```
نیاز نیست property باشه
field باشه. و تو کانستراکتور پر بشه
فرقشون چیست؟

navidnasyry replied:
همه سرویس‌های استفاده شده در کلاس‌ها

navidnasyry replied:
readonly چرا نباشه؟

navidnasyry commented on FullTextSearch/Application/DocumentsReader/Interfaces/IOrDocumentsReader.cs:
```diff
@@ -0,0 +1,6 @@
+
+internal interface IOrDocumentsReader
+{
+    HashSet<string> GetOrDocuments(Dictionary<string, List<string>> invertedIndex, List<string> orWords);
```
read about HashSet
when and where its good practice?

navidnasyry commented on FullTextSearch/Application/DocumentsReader/OrDocumentsReader.cs:
```diff
@@ -0,0 +1,12 @@
+
+internal class OrDocumentsReader : IOrDocumentsReader
+{
+    public HashSet<string> GetOrDocuments(Dictionary<string, List<string>> invertedIndex, List<string> orWords)
```
how about IReadOnlyList<string> orWords?

navidnasyry commented on FullTextSearch/Application/InvertedIndex/InvertedIndexDictionary.cs:
```diff
@@ -0,0 +1,33 @@
+    IInvertedIndexDictionaryFiller invertedIndexDictionaryFiller)
+    : IInvertedIndexDictionary
+{
+    private IQuerySearcher                 Searcher              { get; } = querySearcher;
+    private IInvertedIndexDictionaryFiller IndexDictionaryFiller { get; } = invertedIndexDictionaryFiller;
```
null checker for injected services...

navidnasyry commented on FullTextSearch/Application/InvertedIndex/InvertedIndexDictionaryFiller.cs:
```diff
@@ -0,0 +1,47 @@
+    private Dictionary<string, List<string>> InvertedIndex    { get; } = new();
+
+    public Dictionary<string, List<string>> Build(string filepath)
+    {
+        string[] files = Directory.GetFiles(filepath);
```
var

navidnasyry commented on FullTextSearch/Application/InvertedIndex/QuerySearcher.cs:
```diff
@@ -0,0 +1,73 @@
+internal class QuerySearcher(IWordsProcessor wordsProcessor, ISearcher searcher)
+    : IQuerySearcher
+{
+    private ISearcher                         Searcher       { get; } = searcher;
+    private IWordsProcessor                   WordsProcessor { get; } = wordsProcessor;
+    private Dictionary<string, List<string>>? InvertedIndex  { get; set; }
+    private string[]                          QueryWords     { get; set; } = null!;
+    private bool                              IsConstructed  { get; set; }
```
field

navidnasyry commented on FullTextSearch/Application/InvertedIndex/QuerySearcher.cs:
```diff
@@ -0,0 +1,73 @@
+        return result;
+    }
+
+    private bool AreAllWordTypesPresent() => OrWords.Count > 0 && AndWords.Count > 0;
+    private bool AreAndWordsPresent() => AndWords.Count > 0;
+    private bool AreOrWordsPresent() => OrWords.Count > 0;
```
readonly?

navidnasyry commented on FullTextSearch/Application/Searchers/DataViewModels/Words.cs:
```diff
@@ -0,0 +1,9 @@
+namespace FullTextSearch.Application.Searchers.DataViewModels;
+
+internal class Words
```
read about sealed class

navidnasyry commented on FullTextSearch/Application/Searchers/Interfaces/IAndNotSearcher.cs:
```diff
@@ -0,0 +1,8 @@
+
+internal interface IAndNotSearcher
+{
+    IEnumerable<string> AndNotSearch(Dictionary<string, List<string>> invertedIndex, Words words);
```
how about IReadOnlyList?

navidnasyry commented on FullTextSearch/Application/Searchers/Interfaces/ISearcher.cs:
```diff
@@ -0,0 +1,10 @@
+
+internal interface ISearcher
+{
+    IEnumerable<string> AndOrNotSearch(Dictionary<string, List<string>> invertedIndex, Words words);
+    IEnumerable<string> AndNotSearch(Dictionary<string, List<string>> invertedIndex, Words words);
+    IEnumerable<string> OrNotSearch(Dictionary<string, List<string>> invertedIndex, Words words);
```
same

navidnasyry commented on FullTextSearch/Services.cs:
```diff
@@ -0,0 +1,41 @@
+    {
+        var serviceCollector = new ServiceCollection();
+
+        serviceCollector.AddSingleton<IPorter2Stemmer, EnglishPorter2Stemmer>();
```
What are the advantages of singleton class?
read about Addtransient and Addscope.
When is it dangerous to use singleton classes?

navidnasyry reviewed (changes requested):
excellent :)

Shafagh-Sepehr committed:
refactor: rename all "Interfaces" named folders to "Services"

Shafagh-Sepehr committed:
feat(exceptions): implement public constructors of each exception for handling of exceptions.

Shafagh-Sepehr committed:
refactor(AndDocumentsReader): for intersecting, use foreach loop instead of for loop and extract that part as a method.

Shafagh-Sepehr committed:
refactor: turn all private properties into readonly fields

Shafagh-Sepehr committed:
fix: add null checkers for all injected services.

Shafagh-Sepehr committed:
perf(searchers): change set operations methods to use HashSet methods.

convert linq's Except and Intersect methods to IntersectWith and ExceptWith methods of HashSet<T>.

Shafagh-Sepehr committed:
perf(Words): change Words.cs field types from List to IReadOnlyList.

this affects words-processors which create them and document-readers which use them.

Shafagh-Sepehr committed:
perf(Searchers): change return value types from IEnumerable to IReadOnlySet.

Shafagh-Sepehr committed:
perf(Words): seal Words class to gain performance and preventing inheritance.

Shafagh-Sepehr committed:
perf(AndDocumentsReader): make IntersectAllElements class static.

Shafagh-Sepehr committed:
refactor(StringToWordProcessor): replace long linq expression to several methods for better readability and maintainability.

Shafagh-Sepehr committed:
feat(appsettings): create appsettings.json file for user to maintain custom configurations.

Shafagh-Sepehr committed:
refactor: make all classes sealed because none of them was meant to be inherited
